feat: add pdftk + print_designer + Chromium as pdf renderer

rasos opened the pull request:
fixes "pdftk test call failed" https://github.com/frappe/frappe/issues/27964#issuecomment-2802850845

revant commented:
Why added apps.json and base64 of apps.json?

rasos commented:
This was for our custom build with some additional apps. For a generic app including `print_designer` you would need to add at least the `print_designer` app, that is installing Chrome. 

We then build it with 

```
export APPS_JSON_BASE64=$(base64 -w 0 /path/to/apps.json)

docker build \
  --build-arg=FRAPPE_PATH=https://github.com/frappe/frappe \
  --build-arg=FRAPPE_BRANCH=version-15 \
  --build-arg=PYTHON_VERSION=3.12.8 \
  --build-arg=NODE_VERSION=22.14.0 \
  --build-arg=APPS_JSON_BASE64=$APPS_JSON_BASE64 \
  --tag=registry.osalliance.com/erpnext/erpnextcustom:v15.60.2 \
  --file=images/custom/Containerfile .
```

revant commented:
Remove if it was for you custom build, just send the relevant change

maheini commented:
Thank you for the feedback - I appreciate it. 

Even though wkhtmltopdf has kind of a lot of issues, it's doing it's job in the end. I'd argue it's ok enough for most projects. But it's not on me to decide on that :smiley: 

Pay attention, bench install-app will install chromium, but only within the container it is executed in. If you got a multi-container setup (default compose.yaml), this won't install chromium within the scheduler/queue container.

Therefore, bench install-chrome should be run at build-time of the docker image. I already reported this issue in the print_designer repo [here](https://github.com/frappe/print_designer/issues/432) in order to raise awareness. (I suspect this might be also the case on your current production setup then)

I noticed something else, if I'm not mistaken: you added additional apps like webshop, ecommerce_integration, hrms and non_profit into apps.json, which wouldn't be installed by default and are unrelated to the print_designer specific implementation.

maheini commented:
`bench setup-chrome` stems from print_designer and won't be available without installing print_designer first. You need to install print_designer and it's direct dependencies and then add this command into the Dockerfile in order to get chromium installed at build-time:
```
# Bench commands won't work as root
USER frappe
RUN bench setup-chrome

# Maybe switch back to root if required...
# USER root
```

maheini commented:
This is a bug in print_designer (afaik it's solved in the dev branch) https://github.com/frappe/print_designer/issues/437#issuecomment-2815906587

revant commented:
I'm closing this from here. 

We'll add it when frappe develop branch has it. 

Any other frappe official app, including print designer can have their github actions build it. 

CRM, HRMS, I helped configure. Rest of the new apps are following similar workflow.

rasos commented:
For those who can not wait for Chromium included in frappe, here is a `install_print_designer.sh`  script that installs it for each site: 

```
#!/bin/bash
# Chromium gets lost when pod is restarted, so we install the print_designer app again
# Check if chromium directory is empty
if [ -z "$(ls -A /home/frappe/frappe-bench/chromium)" ]; then
    echo "Chromium directory is empty - installing print_designer on all sites"
    
    # Get all site directories (exclude 'assets' directory)
    sites=$(find /home/frappe/frappe-bench/sites -maxdepth 1 -mindepth 1 -type d -not -name 'assets' -not -name 'archived_sites')
    
    # Install print_designer on each site
    for site_path in $sites; do
        site=$(basename "$site_path")
        echo "Installing print_designer on site: $site"
        bench --site "$site" install-app print_designer --force
    done
else
    echo "Chromium directory not empty - skipping print_designer installation"
fi
```

In the `Containerfile` I added

```
# Copy print_designer app installation script and set permissions https://git.fairkom.net/hosting/k8s-deployments/erpnext/-/issues/7#note_105627
COPY resources/install_print_designer.sh /usr/local/bin/install_print_designer.sh
RUN chmod 755 /usr/local/bin/install_print_designer.sh
```

In the kubernetes deployment I added

```
# install again print designer in gunicorn if chromium got lost after pod restart
          lifecycle:
             postStart:
               exec:
                 command: ["/usr/local/bin/install_print_designer.sh"]
```

So even in case the pod gets moved to another worker node or restarted, it runs the script, and always installs the print_designer app (using --force option), which installs Chromium at every site.
